Support for NRA

marcogario opened the pull request:
Non-linear (polynomial) expressions are now supported by pySMT. This includes the Times, Div and Pow operators.

The only solver currently supporting QF_NRA is z3. For this reason we rely on z3 to represent algebraic solutions (e.g., sqrt(2)). This might be subject to changes in the future, and the constructor ```Algebraic``` is "hidden" in ```FormulaManager```.

This addresses #270.

mikand commented on pysmt/formula.py:
```diff
@@ -231,74 +212,77 @@ def Times(self, left, right):
+            raise ValueError("The exponent of POW must be a constant.", exponent)
 
+        if base.is_constant():
+            val = base.constant_value() ** exponent.constant_value()
```
Why performing this simplification here? In other operators (e.g. Times or Plus) this is not done and is demanded to the simplifier....

marcogario replied:
The idea here (maybe implemented incorrectly) is that constant pow is linear. By doing this simplification here, we do not need to deal with walk_pow in the converters of each solver. A solver that does not support pow, will never see the pow operator if we only have constant pow. Otherwise, we need to implement walk_pow in each converter, and check whether we are in a constant case. This issue does not exist with plus or times, because we still need to handle those operators in all LIRA converters.

Does this make sense?

mikand replied:
OK, let's keep this even if it is not uniform wrt the other operators.

mikand commented on pysmt/numeral.py:
```diff
@@ -0,0 +1,21 @@
+        """Represents a Number (Algebraic)"""
+        def __init__(self, obj):
+            self.obj = obj
+            warnings.warn("Z3 is not installed. We currently do not support stand-alone algebraic numbers. This might lead to unexpected results")
```
Why not a NotImplementedError instead of a warning?

marcogario replied:
It seemed too harsh... You should still be able to get a model, as long as you treat the object as an opaque object. As soon as you try to do anything on it, it will break. This should never happen right now, so I am fine with converting it into ```NotImplementedError```.

mikand replied:
I agree that this should never happen with the code in this PR, so I think an exception is better to signal future issues on this matter.

mikand commented on pysmt/operators.py:
```diff
@@ -55,7 +55,11 @@
-ARRAY_VALUE                                 # Array Value (50)
+ARRAY_VALUE,                                # Array Value (50)
+
+DIV,                                        # Arithmetic Division (51)
```
Is there a reason to explicitly represent Div(a, b) internally instead of using Times(a, Pow(b, -1))? This would reduce the number of operators and simplify the walkers, so unless there is an efficiency or technical reason for having the explicit div I would remove it.

marcogario replied:
I'll try to use the Pow(x, -1) and see what happens. If nothing breaks, I'll remove DIV.

marcogario replied:
Ok, I think this boils down to 2 things:

1) Support for NIA
2) Printing

If we support NIA (that I think is reasonable), then we need to understand what it means to write ```2/x``` and whether we can write ```x^-c``` (negative exponent). From a quick test, it looks like Z3 can solve the first one without problems, while not the second one.

For printing, I think that x/y is more readable than (x * (y ^-1)).

mikand replied:
I tried the followig script (note that the exponentiation operator is ** not ^ ...)
```python
import z3

a = z3.Int("a")
b = z3.Int("b")

c = a / b
d = a * (b ** -1)

f1 = (c == 2)
f2 = (d == 2)

s = z3.Solver()
s.add(f1)
res = s.check()
print("Result for %s: %s" % (f1, res))
if str(res) == "sat":
    print "  Model: " + str(s.model())

print("")
s = z3.Solver()
s.add(f2)
res = s.check()
print("Result for %s: %s" % (f2, res))
if str(res) == "sat":
    print "  Model: " + str(s.model())
```

On my machine this produces:
```
$ python test.py
Result for a/b == 2: sat
  Model: [b = 0, a = 38]

Result for a*b**-1 == 2: unknown
```

So, both seems to be representable in z3, but the second one returns unknown. Moreover, there seems to be a bug, because the model returned by the division yields an invalid model (unless division by 0 is undefined, and therefore free). I think this needs a bit more investigation...

mikand replied:
NOTE

on the website http://rise4fun.com/z3/tutorialcontent/guide you can read the following:

" Z3 also has support for division, integer division, modulo and remainder operators. Internally, they are all mapped to multiplication. "

marcogario replied:
Yes, this is compatible with the tests I made. So, how do you suggest that we proceed? I think that there seems to be a case to leave Div.

Re division by 0: it is undefined (see http://smtlib.cs.uiowa.edu/theories-Reals.shtml)
```Since in SMT-LIB logic all function symbols are interpreted as total
  functions, terms of the form (/ t 0) *are* meaningful in every 
  instance of Reals. However, the declaration imposes no constraints
  on their value. ```

mikand replied:
OK to leave the DIV.

mikand commented on pysmt/printers.py:
```diff
@@ -33,6 +36,34 @@ def __init__(self, stream, env=None):
+        self.set_function(partial(self._walk_nary, " >> "), op.BV_LSHR)
+        self.set_function(partial(self._walk_nary, " a>> "), op.BV_ASHR)
+        self.set_function(partial(self._walk_nary, " bvcomp "), op.BV_COMP)
+
```
:+1:

mikand commented on pysmt/test/examples.py:
```diff
@@ -613,6 +613,31 @@ def get_example_formulae(environment=None):
+            Example(expr=Equals(Pow(p, Int(2)), Int(0)),
+                    is_valid=False,
+                    is_sat=True,
+                    logic=pysmt.logics.QF_NIA),
```
So this PR is about NRA and NIA! :bowtie:

mikand commented on pysmt/test/test_nra.py:
```diff
@@ -0,0 +1,85 @@
+#
```
No test for NIA....

mikand commented:
This looks VERY good! I added few comments in the PR, but generally I love this!

marcogario commented:
Ok. I changed the warning to an NonImplementedError: if this should not be possible, then lets make sure it is not! :)

I renamed the test file in test_nlira, adding (a few) tests for NIA. The only thing missing is adding some files from SMT-LIB.

marcogario commented:
I added examples from SMT-LIB. For me it is ok to merge this. 

@mikand  Can you take a second look and eventually merge?
